app/controllers/users/passwords_controller.rb

```diff
@@ -4,40 +4,23 @@ class PasswordsController < Devise::PasswordsController
 
     before_action :confirm_valid_token, only: [:edit]
 
-    rescue_from Pundit::NotAuthorizedError do |_exception|
-      # We are utilizing Pundit's policy for verifying which user can
-      # recover passwords. However, we always want to return success.
-      flash[:success] = t('notices.password_reset')
-      redirect_to after_sending_reset_password_instructions_path_for(resource_name)
-    end
-
-    # rubocop:disable AbcSize, MethodLength
-    # TODO(sbc): Refactor to address rubocop warnings
     def create
-      resource = resource_class.find_by_email(resource_params[:email])
-
-      if resource
-        authorize resource, :recover_password?
-
-        self.resource = if resource.confirmed_at.nil?
-                          # If the account has not been confirmed, password reset should resend
-                          # the confirmation email instructions
-                          resource_class.send_confirmation_instructions(
-                            resource_params
-                          )
-                        else
-                          # only send_reset_password_instructions if resource is matched above.
-                          # this disallows other roles from using the password recovery form.
-                          resource_class.send_reset_password_instructions(
-                            resource_params
-                          )
-                        end
+      user = User.find_by_email(params[:user][:email]) || NullUser.new
+
```

Comment:
Because that's the not the right logic. We also need to make sure the user's role is `user`. We can combine both conditionals by using Ruby's safe navigation operator, like this:
```ruby
return redirect_with_flash unless user&.role == 'user'
```
which is equivalent to this:
```ruby
return redirect_with_flash unless user && user.role == 'user'
```
However, checking for the presence/absence of a user is basically type checking, which is a smelly pattern in Ruby. We're asking if the User is of type NilClass instead of type User. It violates the "Tell, Don't Ask" principle. Instead, we use the `NullUser`, which is of the correct type, and defines appropriate methods with appropriate values.

So, instead of this:
```ruby
return redirect_with_flash unless user&.role == 'user'
```
We can avoid the type checking by making sure we always get an object of the User type that responds to the `role` method:
```ruby
return redirect_with_flash unless use.role == 'user'
```
We just need to make sure the NullUser has a role that is not set to `user` so that the method can return early. That was the part that I messed up, and I will push a fix for that now.

See these references to learn more about the Null Object pattern and why you should use it:
https://robots.thoughtbot.com/rails-refactoring-example-introduce-null-object
http://www.sandimetz.com/blog/2014/12/19/suspicions-of-nil
https://www.youtube.com/watch?v=OMPfEXIlTVE